pallets/amm/src/lib.rs

```diff
@@ -155,107 +155,123 @@ pub mod pallet {
+            Pools::<T, I>::try_mutate(
+                &base_asset,
+                &quote_asset,
+                |pool_liquidity_amount| -> DispatchResultWithPostInfo {
+                    if let Some(liquidity_amount) = pool_liquidity_amount {
+                        let optimal_quote_amount = Self::quote(
+                            base_amount,
+                            liquidity_amount.base_amount,
+                            liquidity_amount.quote_amount,
+                        );
 
-                let optimal_quote_amount = Self::quote(
-                    base_amount,
-                    pool_liquidity_amount.base_amount,
-                    pool_liquidity_amount.quote_amount,
-                );
+                        let (ideal_base_amount, ideal_quote_amount): (Balance, Balance) =
+                            if optimal_quote_amount <= quote_amount {
+                                (base_amount, optimal_quote_amount)
+                            } else {
+                                let optimal_base_amount = Self::quote(
+                                    quote_amount,
+                                    liquidity_amount.quote_amount,
+                                    liquidity_amount.base_amount,
+                                );
+                                (optimal_base_amount, quote_amount)
+                            };
+
+                        let (minimum_base_amount, minimum_quote_amount) = if is_inverted {
+                            (minimum_amounts.1, minimum_amounts.0)
+                        } else {
+                            (minimum_amounts.0, minimum_amounts.1)
+                        };
+
+                        ensure!(
+                            ideal_base_amount >= minimum_base_amount
+                                && ideal_quote_amount >= minimum_quote_amount
+                                && ideal_base_amount <= base_amount
+                                && ideal_quote_amount <= quote_amount,
+                            Error::<T, I>::NotAIdealPriceRatio
+                        );
 
-                let (ideal_base_amount, ideal_quote_amount): (Balance, Balance) =
-                    if optimal_quote_amount <= quote_amount {
-                        (base_amount, optimal_quote_amount)
-                    } else {
-                        let optimal_base_amount = Self::quote(
-                            quote_amount,
-                            pool_liquidity_amount.quote_amount,
-                            pool_liquidity_amount.base_amount,
+                        let (base_amount, quote_amount) = (ideal_base_amount, ideal_quote_amount);
+                        let ownership = sp_std::cmp::min(
+                            (base_amount.saturating_mul(liquidity_amount.ownership))
+                                .checked_div(liquidity_amount.base_amount)
+                                .expect("cannot overflow with positive divisor; qed"),
```

Comment:
maybe it should fail and just revert back storage instead?

Reply:
@GopherJ Could you please refer any example or hint?

Reply:
I mean you are using `try_mutate` and it accepts `DispatchResult..` maybe `.ok_or(Ari...::Underflow)?` like in loans pallet

Reply:
we don't want `expect` and `unwrap`

Reply:
Done